.changeset/curly-peas-lick.md

```diff
@@ -0,0 +1,5 @@
+---
+'@open-wc/form-helpers': patch
+---
+
+change submit event type to Event for compatibility with jsdom < v21
```

Comment:
This is a type change which would count as breaking as someone who leveraged the expected type here previously could be forced to alter their code to support this change.  

If `SubmitEvent` is the event type that belongs here, then it's likely better to keep dispatching that type. If it's missing in contexts like `jsdom@<21` then it would be better to suggest patching that context by adding the EventType to the global scope or enforcing the use of >=21. I see the appeal, but better to be true to the spec and let consumption contexts work it out than possibly make code that could have difficulty working when swapping between the polyfill and not...

Reply:
i get where you’re coming from but not really sure what to do. in jsdom 20, SubmitEvent isn’t implemented, so it’s not newable and this code fails. so in order for jest/jsdom to work, consumers would be required to npm override jsdom 21. i was hoping that SubmitEvent and Event are functionally synonymous

Reply:
maybe instead of updating try to push hard for an update package to jest-environment-jsdom that depends on jsdom 21?

Reply:
A couple things. First we're still in v0 (which should change ASAP as far as I'm concerned), so breaking changes are fine-ish. Second, we can always do something like this:

```typescript
const SubmitEvent = typeof globalThis.SubmitEvent !== 'undefined' ? SubmitEvent : Event;

/** ... */

new SubmitEvent()
```

Reply:
Unless this is set to global it still prevents event callbacks like:
```
function handleSubmit(event: SubmitEvent): void {
    if (event instanceOf SubmitEvent) {
        // do stuff it's a submit event
    }
}
```

Is a little polyfilling is warranted then maybe:
```
globalThis.SubmitEvent = typeof globalThis.SubmitEvent !== 'undefined' ? SubmitEvent : Event;
```

Not sure that fully covers the JSdoms of the world or not.

Reply:
@michaelwarren1106 thoughts on this one? I'm OK with it … 

Reply:
yep, i’ll add that after kiddos bedtime

Reply:
feels like i shouldn’t do this polyfill directly in the submit helper method itself right? feels like i should do it in some index file. but then that would need to be marked as having side-effects for bundlers to not throw it out. thoughts on where best to put this line of polyfill?

Reply:
It’s not a whole lot of work…doing it in that method or doing it in the file next to the method (if you set it to a variable and use that instead of the global it shouldn’t shake, right?) it shouldn’t have a measurable cost.

Reply:
ya, i kinda meant that since it’s a polyfill it prolly shouldn’t change a global in the middle of execution? imo it should change it on import so that it’s changed for the whole runtime instead of just when that submit method is called for the first time?

Reply:
Makes sense. You _should_ be able to get that via something like:

```
const PolyfilledSubmitEvent = globalThis.SubmitEvent = typeof globalThis.SubmitEvent !== 'undefined' ? SubmitEvent : Event;

export const submit = (form: HTMLFormElement): void => {
  if (!form.noValidate && !form.reportValidity()) {
    return;
  } else {
    const submitEvent = new PolyfilledSubmitEvent('submit', {
      bubbles: true,
      cancelable: true
    });
    form.dispatchEvent(submitEvent);
    if (!submitEvent.defaultPrevented) {
      form.submit();
    }
  }
};
```
The usage of `PolyfilledSubmitEvent` within the method should prevent the code from being effected by the tree shake, and the variable setting outside of the method should polyfill at runtime? I'm sure there's something I'm missing, but it feels right...

Reply:
i’ll give it a go! tyvm

Reply:
and thanks again for the valuable feedback! i think these pkgs are super close to a 1.0.0 as caleb said

Reply:
I'm always surprised you haven't flipped the switch already, but it's cool to see all this nuance getting ironed out in the meantime!

Reply:
i’ve been on a quest to get things working in jest/jsdom as well as i can.

Reply:
thoughts on this?

https://github.com/open-wc/form-participation/pull/49/commits/6a8753ab0115b3d7a73d5af9fa89ade1959fd37e#diff-f564f1a07d910e08e0cfaebf61151d27dc0efef9a37cf77d0dc7e71365ebf6cdR7